types/scope.d.ts

```diff
@@ -0,0 +1,115 @@
+// Copyright IBM Corp. 2018. All Rights Reserved.
+// Node module: loopback-datasource-juggler
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+// tslint:disable:no-any
+
+import {AnyObject, Options, Callback, PromiseOrVoid} from './common';
+import {ModelClass, ModelData, ModelBase} from './model';
+import {Count} from './persisted-model';
+import {Filter, Where} from './query';
+
+/**
+ * Definition metadata for scopes
+ */
+export declare class ScopeDefinition {
+  constructor(definition: AnyObject);
+
+  isStatic: boolean;
+  modelFrom?: ModelClass;
+  modelTo?: ModelClass;
+  name: string;
+  params?: AnyObject;
+  methods?: AnyObject<Function>;
+  options?: Options;
+
+  targetModel(receiver: ModelClass | (() => ModelClass)): void;
+
+  /**
+   * Find related model instances
+   * @param {*} receiver The target model class/prototype
+   * @param {Object|Function} scopeParams
+   * @param {Boolean|Object} [condOrRefresh] true for refresh or object as a filter
+   * @param {Object} [options]
+   * @param {Function} cb
+   * @returns {*}
+   */
+  related(
+    receiver: ModelClass | (() => ModelClass),
+    scopeParams: AnyObject | (() => AnyObject),
+    condOrRefresh: boolean | AnyObject,
+    options?: Options,
+    callback?: Callback,
+  ): PromiseOrVoid;
+
+  /**
+   * Define a scope method
+   * @param {String} name of the method
+   * @param {Function} function to define
+   */
+  defineMethod(name: string, fn: Function): Function;
+}
+
+/**
+ * Define a scope to the class
+ * @param {Model} cls The class where the scope method is added
+ * @param {Model} targetClass The class that a query to run against
+ * @param {String} name The name of the scope
+ * @param {Object|Function} params The parameters object for the query or a function
+ * to return the query object
+ * @param methods An object of methods keyed by the method name to be bound to the class
+ */
+export declare function defineScope(
+  cls: ModelClass,
+  targetClass: ModelClass,
+  name: string,
+  params: AnyObject,
+  methods: AnyObject<Function>,
+  options?: Options,
+): ScopeDefinition;
+
+/**
+ * Methods injected by scopes
+ */
+export interface ScopedMethods<T extends ModelBase = ModelBase> {
+  build(data: ModelData<T>): T;
+
+  create(
```

Comment:
@zbarbuto Thank you for chiming in. It's a good idea to use overloaded method signatures. Let's do it incrementally. I would like to use the type information for LoopBack 4 soon.

For `loopback` module, I think there is already a community module - `@types/loopback`. Do you want to check out?

Reply:
Yes have used this in the past - it does have some issues. Was only mentioning in case it was on the cards to eventually have typedefs out of the box in `loopback` as well.

Understood re: changing incrementally.

Reply:
@zbarbuto I don't think I include `Model` in the declaration files. Please note that `ModelBase` is different from `Model`:

- ModelBase: a base class in `loopback-datasource-juggler` as the common root of all models
- Model: a class in `loopback`